### ISSUE TITLE
Large message size.

### ISSUE THREAD

**Issue by leo12chandu** [akkadotnet/akka.net#3295]
When creating a new issue, please make sure the following information is part of your issue description. (if applicable). Thank You!

- Which Akka.Net version you are using
1.3.3
- On which platform you are using Akka.Net
Windows
- A list of steps to reproduce the issue. Or an gist or github repo which can be easily used to reproduce your case.
We will have large messages exchanging within our actor system. I know its recommended to use immutable messages. But if we end up building our large message type as immutable, that would mean the message gets serialized and deserialized it goes through different actors. We do eventually break down into smaller tasks, run em in a pool and aggregate the results back. But curious if there is a way to pass around references atleast within a single node on a single machine within the cluster. I understand if cluster spans multiple nodes and multiple machines, we have to serialize and deserialize.

Another part of the question is also if there is patterns to handle aggregating the results from these small jobs ran by worker nodes back to a single result we send back to the user. essentially we creating a single stateful aggregator to cache this info. As small jobs finish, they send a message to aggregator and we keep track of its state. Once done, we send one response back. But with a single aggregator it causes "bottle neck" and "single point of failure". I was wondering if there a way to avoid this.

**Comment by vasily-kirichenko** [comment]
@leo12chandu Im new to akka, but the aggregation problem is a good job for DistributedData. Or maybe just spawn an fresh actor for each aggregate and send it’s IActorRef with tasks to workers?

**Comment by Horusiath** [comment]
@leo12chandu usually the recommended way of working with large messages is to:

1. If possible collocate actors on the same machine to avoid passing big data chunks over the wire. Standard routers can do this. Clustering helps here in situation, when you have multiple big files to process - then each node will focus on a single one.
2. If you need to pass big message to another node, it's better to use dedicated side channel, like database, file system or even FTP for that. The reasons for this are: 
    - Not to stretch memory buffer without need - if you want to serialize big message, you need to fit it into memory first, what if you'll be processing many messages like that in paralllel?
    - Since there's a single communication channel between every two actor systems, passing big chunk of data will effectively throttle other messages from being send. This will be irritating for small messages, which will be delivered with higher latency. But most of all it may we dangerous, when time-sensitive messages - like akka cluster heartbeats and control events - will not reach their destination on time.

**Comment by leo12chandu** [comment]
@vasily-kirichenko - That is really interesting what you are suggesting. I am new to akka.net myself. How do you spin up fresh actors on demand? I thought we have to specify the number of instances in the config and thats all the number of actors you get until you go back and change the config. Also if I used DistributedData, wouldnt it be the same issue of having to write the large message into DistData and then read it back in the next actor? Thats as bad as serializing everytime as a message and deserializing back in the next actor right?

**Comment by leo12chandu** [comment]
@Horusiath - Thank you. Btw, Love your article on Don't as Tell. Infact we are using the combination of Forward and Aggregator pattern in our application. 
1. If we collocate the actors on a single node/machine, does it mean it does not serialize/deserialize everytime and just pass the large message by reference? Or are you saying this just avoids over the network latency? I've tried setting up a collocated node cluster and when I sent a message from outside using ClusterClient, I see that it still tried to serialize/deserialize But when I  tried to send a message within cluster from service start, it did not serialize/deserialize. Which is odd to me.

2. Got it. That is interesting to know that there is single communicate channel between every two actor systems. Which means we may have to make sure we design our system in such a way that we use 1 actor system for all to avoid a single channel. 

Thanks,
Chandra.

**Comment by leo12chandu** [comment]
@vasily-kirichenko - I realized my mistake about fresh actors. Ofcourse I can spin up new ones from the Receive method. That was so silly of me. I am going to edit my message and get rid of that question..

**Comment by leo12chandu** [comment]
@vasily-kirichenko - I've used the fresh actor-per-request and that solved the issue of having to keep track of the requests! Thanks.

Now I have to figure out how to collocate all actors that belong to a node to always serve requests by delegating to actors within that node/machine as opposed to sending em to a new node/machine using the routers.

**Comment by leo12chandu** [comment]
I was able to achieve collocation in a way using roles. Let me know if I went on a tangent. I had 1 node with all the actors necessary and assigned a role to that node (role1). I had second node with all the same actors now running with a different role (role2). This allowed requests going to node1’s root actor to only use actors in that node1. similarly node2’s requests are all handled internally. I could extend by creating more nodes obviously. 

May be even if I removed roles completely I might get the same affect? But I dont want these deployed into lighthouse/seed node. hence roles are important to maintain.

**Comment by boopalanjayaraman** [comment]
I'm dealing with a similar financial aggregation problem. I'd like to send pieces of a big report for processing into different child actors that are co-located so that we can avoid serializing and deserializing them repeatedly. We'd want the solution to be scalable horizontally as well.

Assigning a Role is the only way of doing it? Or do we have other means? 

I have another basic question, the big report is formed into a kind of a tree (which should be aggregated from the bottom of tree to way up, for ex using post-order depth first traversal), and my intention is to pass the branches to different actors so the calculations happen in parallel. But how do I let the actors aggregate and modify the values of different objects if I pass only read-only properties (in order to be immutable)?

**Comment by leo12chandu** [comment]
If you have to parallelize work and have all your nodes/actors running
within the same actor system, using roles is a common way of having to
avoid serialization/deserialization to keep large messages from travelling
across the wire through several nodes. If you dont have multiple nodes,
obviously you dont need to use roles. But I imagine you want scalability,
hence you need to use multiple nodes in clustering setup. You also have
certain pooling strategies like hashing to route messages to different
actors, but that gets a little more complex setup.

About your tree, if you need to process it as an infix to postfix notation,
you will have to follow something to the extent of shunting yard algorithm
<https://en.wikipedia.org/wiki/Shunting-yard_algorithm> broadly, but I am
digressing from the topic. You could have potentially multiple child actors
run sub aggregations and kick back the results to main aggregator actor as
you had stated. You are right that the main aggregator has to in turn
extract the results out of these immutable subaggregation results if
messages are immutable. The reason for immutability is to avoid stupidity
with certain folks meddling with mutable objects in multiple threads
asynchronously and not realizing the state of the mutable objects had
changed (remember dispatcher behind the akka system). If you dont want to
have to bear the cost of reconstructing the results out of immutable
objects (which may not be a big deal except slight CPU hit), you could
potentially make your messages mutable. However mutable messages are
definitely not recommended. So, as a solutions architect its the cost you
have to weigh on your end.
